streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/HighAvailabilityTaskAssignorTest.java

```diff
@@ -0,0 +1,732 @@
+    }
+
+    @Test
+    public void shouldAssignStandbysForStatefulTasks() {
+        numStandbyReplicas = 1;
+        allTasks = mkSet(task0_0, task0_1);
+        statefulTasks = mkSet(task0_0, task0_1);
+
+        client1 = getMockClientWithPreviousCaughtUpTasks(mkSet(task0_0));
+        client2 = getMockClientWithPreviousCaughtUpTasks(mkSet(task0_1));
+
+        clientStates = getClientStatesWithTwoClients();
+        createTaskAssignor();
+        taskAssignor.assign();
+
+        assertThat(client1.activeTasks(), equalTo(mkSet(task0_0)));
+        assertThat(client2.activeTasks(), equalTo(mkSet(task0_1)));
+        assertThat(client1.standbyTasks(), equalTo(mkSet(task0_1)));
+        assertThat(client2.standbyTasks(), equalTo(mkSet(task0_0)));
+    }
+
+    @Test
+    public void shouldNotAssignStandbysForStatelessTasks() {
+        numStandbyReplicas = 1;
+        allTasks = mkSet(task0_0, task0_1);
+        statefulTasks = emptyTasks;
+
+        client1 = getMockClientWithPreviousCaughtUpTasks(emptyTasks);
+        client2 = getMockClientWithPreviousCaughtUpTasks(emptyTasks);
+
+        clientStates = getClientStatesWithTwoClients();
+        createTaskAssignor();
+        taskAssignor.assign();
+
+        assertThat(client1.activeTaskCount(), equalTo(1));
+        assertThat(client2.activeTaskCount(), equalTo(1));
+        assertHasNoStandbyTasks(client1, client2);
+    }
+
+    @Test
+    public void shouldAssignWarmupReplicasEvenIfNoStandbyReplicasConfigured() {
+        allTasks = mkSet(task0_0, task0_1);
+        statefulTasks = mkSet(task0_0, task0_1);
+        client1 = getMockClientWithPreviousCaughtUpTasks(mkSet(task0_0, task0_1));
+        client2 = getMockClientWithPreviousCaughtUpTasks(emptyTasks);
+
+        clientStates = getClientStatesWithTwoClients();
+        createTaskAssignor();
+        taskAssignor.assign();
+        
+        assertThat(client1.activeTasks(), equalTo(mkSet(task0_0, task0_1)));
+        assertThat(client2.standbyTaskCount(), equalTo(1));
+        assertHasNoStandbyTasks(client1);
+        assertHasNoActiveTasks(client2);
+    }
+
+    @Test
+    public void shouldNotAssignMoreThanMaxWarmupReplicas() {
+        maxWarmupReplicas = 1;
+        allTasks = mkSet(task0_0, task0_1, task0_2, task0_3);
+        statefulTasks = mkSet(task0_0, task0_1, task0_2, task0_3);
+        client1 = getMockClientWithPreviousCaughtUpTasks(mkSet(task0_0, task0_1, task0_2, task0_3));
+        client2 = getMockClientWithPreviousCaughtUpTasks(emptyTasks);
+
+        clientStates = getClientStatesWithTwoClients();
+        createTaskAssignor();
+        taskAssignor.assign();
+
+        assertThat(client1.activeTasks(), equalTo(mkSet(task0_0, task0_1, task0_2, task0_3)));
+        assertThat(client2.standbyTaskCount(), equalTo(1));
+        assertHasNoStandbyTasks(client1);
+        assertHasNoActiveTasks(client2);
+    }
+
+    @Test
+    public void shouldAssignStandbyReplicasInAdditionToWarmupReplicas() {
```

Comment:
```suggestion
    public void shouldReUseWarmupReplicasForStandbyReplicas() {
```

Just a name change suggestion. The behavior in the test is what I was expecting, but the method name led me to believe that we were basically doubling up on replicas (ie we'd get three copies of the replica: an active, a warmup, and a standby).

Reply:
Does `shouldNotAssignWarmupAndStandbyToTheSameClient` make sense to you? I would interpret `ReuseWarmupForStandby` as meaning the warmup replicas will count as a standby (in terms of `num.standby.replicas` ), which is not the case.